### PR TITLE
Squashed commit of the following:

### PR DESCRIPTION
**Pull request by gonzaloiglesiasiglesias**
New Rule Extraction Algorithm:
- Simplification of existing logic
- A new compat_mode, which mostly replicates existing logic
- Unit and Integration tests
- Written in scala

Command Line Interface (CLI):
-The CLI parser now accepts unknown options. This means the options can be placed in one big config file and shared between tools.
-At Adrià’s request you only need to specify provenances once when running MapReduce jobs and the features argument is not needed. In the example README I only specify the config for rule loading and extraction. The phrase probability (S2T, T2S) jobs and merging do not need any config now. The mapreduce_features argument is now dead.
-Command line arguments are automatically added to the Hadoop config, for later referral by MapReduce jobs.
-Killed the rule filtering config file. The parameters at the beginning of the file have been moved to the general CLI, and can be placed in config using the ‘@‘ syntax. The rule patterns have been moved to separate files.
-Killed the CF.rulextract.patterns config file. Instead the generation uses source patterns passed on the filtering config.

Rule Filtering:
-Killed skip patterns
-Two files containing rule patterns: one containing allowed patterns, and one containing the source pattern constraints
-Rewrote the rule filtering logic so that it is more understandable.
-The result is around 3000 new rules are added to the set specific grammar, probably due to the omission of skip patterns. This is a 0.3% increase in the number of rules. The rest of the set specific grammar is exactly the same.
-Filtering is applied twice, at merging and at retrieval. This results in smaller hfiles but still allows for experimentation with more generous thresholds.

Adding Features:
- Have rewritten the feature creation logic.
- A simpler feature registration mechanism:
	- Add an enum to represent the feature, which is one line.
	- Write a function to compute the feature. The function is given the phrase probabilities, the rule counts, and the alignments. The output is an array of doubles representing the feature values. I think this is enough to compute the BBN traits.
	- Register the function, which is again one line.
- Most features are now modelled by defaults. For example the rule insertion penalty (RIP) is 1 for every rule. We don’t need to compute this with a function.
- Glue, deletion, and OOV rules are completely modelled by defaults. We do not do any dynamic feature generation on these types of rules.

General code cleanup:
- Many string literals were repeated defined throughout the codebase. This is just plain wrong in a statically typed language (Java, C++, etc). Define the string literal once in a constant, and then refer to symbol of the constant.
- Deleted repeated code and repeated logic checks.
- Decoupled the lexical servers and retrieval from Hadoop. They have nothing to do with MapReduce and HDFS so they don’t need dependencies on it.

### CODE REVIEW

File: java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		trainingDataFile = copyDataToTestDir(TRAINING_DATA);
+	}
+	
+	@AfterClass
+	public static void cleanUp() throws IOException{
+		folder.delete();
+	}
+
+	
+	private boolean isContigious(List<Rule> rules, Function<Rule, List<Symbol>> getStr){
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java]: isContiguous?

File: java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import org.junit.rules.TemporaryFolder;
+
+import uk.ac.cam.eng.extraction.hadoop.datatypes.TextArrayWritable;
+import uk.ac.cam.eng.extraction.hadoop.features.phrase.Source2TargetJob;
+import uk.ac.cam.eng.extraction.hadoop.features.phrase.Target2SourceJob;
+import uk.ac.cam.eng.util.Pair;
+
+/**
+ * @author aaw35
+ *
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java]: General request. Please add documentation to the code. It is not necessary to add per method, just one per class or at least per file, describing in some detail what is done.

File: java/ruleXtract/src/it/resources/CF.rulextract.patterns [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# source pattern constraints
+# nocc is the minimum number of occurrences of a hierarchical
+# rule in the training data
+# ntrans is the maximum number of translations per source
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/resources/CF.rulextract.patterns]: nocc and ntrans are not explicitly used in this config, so these comments are somewhat odd. You could put commented line showing usage examples

File: java/ruleXtract/src/it/scala/uk/ac/cam/eng/PipelineTest.scala [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import uk.ac.cam.eng.extraction.hadoop.merge.MergeJob.MergeFeatureMapper
+import uk.ac.cam.eng.extraction.hadoop.merge.MergeJob.MergeRuleMapper
+import uk.ac.cam.eng.rule.features.Feature
+import uk.ac.cam.eng.rule.retrieval.HFileRuleReader
+import uk.ac.cam.eng.util.CLI
+import uk.ac.cam.eng.util.CLI.FilterParams
+import uk.ac.cam.eng.extraction.hadoop.datatypes.FeatureMap
+import uk.ac.cam.eng.extraction.hadoop.datatypes.ExtractedData
+
+class PipelineTest extends FunSuite with BeforeAndAfterAll {
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/scala/uk/ac/cam/eng/PipelineTest.scala]: Please document classes in some detail. What is PipelineTest supposed to do? Is this testing the full pipeline?

File: java/ruleXtract/src/it/scala/uk/ac/cam/eng/extraction/ExtractIntegrationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import scala.collection.mutable.ArrayBuffer
+import org.scalatest._
+import org.apache.hadoop.io.SequenceFile
+import org.apache.hadoop.fs.FileSystem
+import org.apache.hadoop.io.MapWritable
+import uk.ac.cam.eng.extraction.hadoop.datatypes.TextArrayWritable
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.io.Text
+
+
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/scala/uk/ac/cam/eng/extraction/ExtractIntegrationSpec.scala]: Please add documentation

File: java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java (hunk 2 above)
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/java/uk/ac/cam/eng/extraction/RuleExtractorTest.java]: If that documentation is readable by doxygen, that would be great.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/AlignmentCountMapWritable.java [hunk 6]
```diff
@@ -20,20 +20,27 @@
...
-import java.util.Map.Entry;
 
 import org.apache.hadoop.io.Writable;
 import org.apache.hadoop.io.file.tfile.Utils;
 
+import uk.ac.cam.eng.extraction.Alignment;
+
 /**
  * @author Juan Pino
  * @date 14 July 2014
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/AlignmentCountMapWritable.java]: Suggest to add yourself here as author
Also please document this file/classes

File: java/ruleXtract/scripts/retrieve.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+DIR=`dirname $0`
+MEM=${RETRIEVE_MEM:-70G}
+SCALA=${SCALA_BIN:-scala}
+echo "Using ruleXtract jar: $DIR/../target/ruleXtract.jar"
+exec $SCALA -nc -J-Xms"$MEM" -J-Xmx"$MEM" -classpath "$DIR/../target/ruleXtract.jar" $0 $@
+!#
+
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/scripts/retrieve.scala]: Could you add a small blurb here to explain what is this script doing?

File: java/ruleXtract/src/it/scala/uk/ac/cam/eng/PipelineTest.scala [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    FileInputFormat.setInputPaths(t2sJob, extractOut)
+    FileOutputFormat.setOutputPath(t2sJob, t2sOut);
+    val ft2s = Future {
+      t2sJob.waitForCompletion(true)
+    }
+    Await.ready(fs2t, 1 hours)
+    Await.ready(ft2s, 1 hours)
+
+    conf.set(FilterParams.MIN_SOURCE2TARGET_PHRASE, "0.01");
+    conf.set(FilterParams.MIN_TARGET2SOURCE_PHRASE, "1e-10");
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/it/scala/uk/ac/cam/eng/PipelineTest.scala]: It looks like this is your script for integration testing that takes 5 minutes long after reducing the size of the parallel data. If you want to increase data coverage without hitting the time, you could try modifying FilterParams.MIN_TARGET2SOURCE_PHRASE to 0.01 here (same for FilterParams.MIN_TARGET2SOURCE_RULE). This might reduce considerably the number of extracted rules.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/ProvenanceProbMap.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+		return instance.size();
+	}
+
+	public boolean isEmpty() {
+		return instance.isEmpty();
+	}
+
+	public boolean containsKey(Object key) {
+		return instance.containsKey(key);
+	}
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/ProvenanceProbMap.java]: Please document this file/class

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/ProvenanceProbMap.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+import org.apache.hadoop.io.DoubleWritable;
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.io.WritableUtils;
+
+/**
+ * 
+ * @author Aurelien Waite
+ * @date 28 May 2014
+ */
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/ProvenanceProbMap.java]: Please document this file.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/RuleData.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+import org.apache.hadoop.io.Writable;
+
+
+/**
+ * @author Aurelien Waite
+ * @date 7 November 2014
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/RuleData.java]: Please document this file.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/TargetFeatureList.java [hunk 12]
```diff
@@ -36,121 +32,15 @@
  * @author Juan Pino
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/datatypes/TargetFeatureList.java]: Please document

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java [hunk 13]
```diff
@@ -150,86 +161,44 @@ protected void map(MapWritable key, TextArrayWritable value,
 						ruleInfo.putProvenanceCount(prov2Id.get(prov), ONE);
 					}
 				}
-				ruleInfo.putAlignmentCount(aw, 1);
-				context.write(rw, ruleInfo);
+				ruleInfo.putAlignmentCount(ra.getSecond(), 1);
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java]: Should you be using ONE instead of 1?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java [hunk 14]
```diff
@@ -150,86 +161,44 @@ protected void map(MapWritable key, TextArrayWritable value,
...
-		public String max_terminal_length = "5";
-
-		@Parameter(names = { "--max_nonterminal_span" }, description = "Maximum number of source terminals covered by a right-hand-side source nonterminal in a hiero rule")
-		public String max_nonterminal_span = "10";
-
-		@Parameter(names = { "--provenance" }, description = "Comma-separated provenances")
-		public String provenance;
-	}
-
 	public int run(String[] args) throws FileNotFoundException, IOException,
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java]: Please document the meaning of 0,1 for this method. Also, perhaps using constants would be better (RUN_SUCCESS, RUN_FAILURE, ...)

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/lexical/TTableServer.java [hunk 15]
```diff
@@ -269,73 +264,18 @@ public void close() throws IOException {
...
-			cmd.usage();
+		CLI.TTableServerParameters params = new CLI.TTableServerParameters();
+		Util.parseCommandLine(args, params);
+
+		try (TTableServer server = new TTableServer()) {
+			server.setup(params);
+			server.startServer();
+			System.err.println("TTable server ready on port: "
+					+ server.serverSocket.getLocalPort());
+			Thread.sleep(24 * 60 * 60 * 1000); // Sleep for 24 hours
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/lexical/TTableServer.java]: Does this mean that the server will be 24 hours available tops and then it will shut down? Is this regardless of the server activity? Could you please document here the behaviour?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/phrase/PhraseJob.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+			Util.parseCommandLine(args, params);
+		} catch (ParameterException e) {
+			return 1;
+		}
+		Configuration conf = getConf();
+		Util.ApplyConf(params, conf);
+		Job job = getJob(conf);
+		FileInputFormat.setInputPaths(job, params.input);
+		FileOutputFormat.setOutputPath(job, new Path(params.output));
+		return job.waitForCompletion(true) ? 0 : 1;
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/phrase/PhraseJob.java]: These 0,1 values seem to be the same return values used in other places. You could specify constants instead.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/merge/MergeJob.java [hunk 17]
```diff
@@ -147,60 +190,40 @@ public static Job getJob(Configuration conf) throws IOException {
...
+			MultipleInputs.addInputPath(job, featurePath,
+					SequenceFileInputFormat.class, MergeFeatureMapper.class);
 		}
+		Path rulePath = new Path(params.inputRules);
+		MultipleInputs.addInputPath(job, rulePath,
+				SequenceFileInputFormat.class, MergeRuleMapper.class);
 
-		return 1;
+		FileOutputFormat.setOutputPath(job, new Path(params.output));
+		return job.waitForCompletion(true) ? 0 : 1;
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/merge/MergeJob.java]: See my comment on constants

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/CopyRecordsForTesting.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+ * 
+ * @author aaw35
+ *
+ */
+
+public class CopyRecordsForTesting {
+
+	public static void main(String[] args) throws IOException {
+		if (args.length != 3) {
+			System.err
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/CopyRecordsForTesting.java]: Odd indentation.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java [hunk 19]
```diff
@@ -42,27 +44,27 @@
...
 		final Configuration conf = job.getConfiguration();
 		Path file = getDefaultWorkFile(job, ".hfile");
 		FileSystem fs = file.getFileSystem(conf);
-		HFile.WriterFactory writerFactory = HFile.getWriterFactory(conf);
-		final HFile.Writer writer = writerFactory.createWriter(fs, file,
-				64 * 1024, "gz", null);
 		final CacheConfig cacheConfig = new CacheConfig(conf);
-		return new RecordWriter<Text, TargetFeatureList>() {
+		HFile.WriterFactory writerFactory = HFile.getWriterFactory(conf, cacheConfig).withPath(fs, file)
+				.withBlockSize(64 * 1024).withCompression("gz");
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java]: Please document 64 *1024

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java [hunk 20]
```diff
@@ -42,27 +44,27 @@
...
+		final HFile.Writer writer = writerFactory.create();
+		return new RecordWriter<RuleString, TargetFeatureList>() {
 
 			private ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
 
 			private DataOutputStream out = new DataOutputStream(bytesOut);
 
 			BloomFilterWriter bloomFilterWriter = BloomFilterFactory
-					.createBloomAtWrite(conf, cacheConfig, BloomType.ROW, -1,
+					.createGeneralBloomAtWrite(conf, cacheConfig, BloomType.ROW, -1,
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java]: What's the meaning of -1?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+		minSource2TargetRuleLog = Math.log(params.minSource2TargetRule);
+		minTarget2SourceRuleLog = Math.log(params.minTarget2SourceRule);
+		loadConfig(params.allowedPatternsFile, conf,
+				line -> allowedPatterns.add(RulePattern.parsePattern(line)));
+		loadConfig(params.sourcePatterns, conf,
+				line -> {
+					String[] parts = line.split(" ");
+					if (parts.length != 3) {
+						throw new RuntimeException(
+								"line should have 3 fields: " + line);
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java]: Please specify what these 3 fields are.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+			// in case of ties we either keep or don't keep the ties
+			// depending on the config
+			RulePattern rulePattern = RulePattern.getPattern(entry.getFirst());
+			int count = (int) entry.getSecond().getProvCounts().get(countIndex)
+					.get();
+			boolean notTied = count != prevCount;
+			boolean moreThan1NT = sourcePattern.hasMoreThan1NT();
+			if (notTied
+					&& ((moreThan1NT
+							&& nTransConstraint <= numberTranslationsMonotone && nTransConstraint <= numberTranslationsInvert) || (!moreThan1NT && nTransConstraint <= numberTranslations))) {
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java]: weird indentation

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+				}
+			}
+			numberTranslations++;
+			prevCount = count;
+		}
+		return results;
+	}
+
+	/**
+	 * Indicates whether we should filter a rule. The decision is based on a
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java]: returns true if we should filter a rule, or false otherwise

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+	 * particular provenance.
+	 * 
+	 * @return
+	 */
+	private boolean filterRule(Feature s2t, Feature t2s,
+			SidePattern sourcePattern, Rule rule, RuleData data, int provMapping) {
+		IntWritable provIW = IntWritableCache.createIntWritable(provMapping);
+		// Immediately filter if there is data for this rule under this
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java]: Should this comment say "if there is no data" ?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/HFileRuleQuery.java [hunk 25]
```diff
@@ -149,56 +136,38 @@ public void run() {
...
 				}
-
 			}
 			drainQueue();
 		} catch (IOException e) {
 			e.printStackTrace();
 			System.exit(1);
 		}
+		
 		System.out
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/HFileRuleQuery.java]: Weird indentation. This can be in one line.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/RuleRetriever.java [hunk 26]
```diff
@@ -378,90 +312,86 @@ public boolean accept(File dir, String name) {
...
+			retriever.writeGlueRules(out);
+		}
+		System.out.println(retriever.foundPassThroughRules);
+		System.out.println(retriever.foundTestVocab);
+		if (retriever.targetVocabFile != null) {
+			try (BufferedWriter out = new BufferedWriter(new FileWriter(
+					retriever.targetVocabFile))) {
+				for (Set<Symbol> words : retriever.targetSideVocab.subList(1,
+						retriever.targetSideVocab.size())) {
+					out.write("1 2");
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/RuleRetriever.java]: What's the meaning of "1 2"?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/RuleRetriever.java [hunk 27]
```diff
@@ -378,90 +312,86 @@ public boolean accept(File dir, String name) {
...
-							.contains(asciiConstraint)) {
-						features.writeRule(asciiConstraint,
-								AlignmentAndFeatureMap.EMPTY,
-								EnumRuleType.ASCII_OOV_DELETE, out);
-					}
+			}
+			// Add Deletion and OOV rules
+			Rule deletionRuleWritable = new Rule();
+			RuleString dr = new RuleString();
+			dr.add(Symbol.deserialise(-7));
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/RuleRetriever.java]: Is -7 and -6 an integer representation for DR and OOV? Shouldn't they be defined as constants somewhere?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/util/CLI.java [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+ * limitations under the License.
+ *
+ *******************************************************************************/
+
+package uk.ac.cam.eng.util;
+
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.Parameters;
+import com.beust.jcommander.ParametersDelegate;
+
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/util/CLI.java]: Please document this file and add yourself as author.

File: java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala [hunk 29]
```diff
@@ -0,0 +1,276 @@
...
+  def extractPhrasePairs(options: ExtractOptions)(source: Seq[Symbol], target: Seq[Symbol], alignment: Alignment): (Seq[(Span, Span)], Seq[(Span, Span)], Seq[(Span, Span)]) = {
+    val results = new ListBuffer[(Span, Span)]
+    for (i <- 0 to source.length) {
+      var minTargetIndex = Int.MaxValue
+      var maxTargetIndex = Int.MinValue
+      if (alignment.isSourceAligned(i)) {
+        minTargetIndex = math.min(alignment.getS2T(i).head, minTargetIndex)
+        maxTargetIndex = math.max(alignment.getS2T(i).last, maxTargetIndex)
+      }
+      //For heiro we consider phrases with max length of the sentence. The long phrases are filtered out at the end.
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala]: hiero

File: java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+      Tuple2(TwoNTSpan(srcSpan.endX, srcSpan.end, srcSpan.endX, srcSpan.endX, srcSpan.startX2, srcSpan.endX2), true)
+    else if (!options.compatibilityMode && srcSpan.end == srcSpan.endX2)
+      Tuple2(TwoNTSpan(srcSpan.start, srcSpan.startX2, srcSpan.startX,
+        srcSpan.endX, srcSpan.startX2, srcSpan.startX2), true)
+    else
+      (srcSpan, false)
+
+  def add(opts: ExtractOptions, results: Buffer[(Span, Span)], srcSpan: Span, trgSpan: Span,
+          dedupe: Boolean, deduper: scala.collection.mutable.HashSet[Span]): Unit =
+    if (filterPassNonTerminalRule(opts, srcSpan) && (if (dedupe && opts.removeMonotonicRepeats) deduper add srcSpan else true))
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala]: would prefer to avoid these super long scala lines, if possible.

File: java/ruleXtract/src/test/resources/TestConfigFile [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+--ttable_s2t_server_port=4949
+--ttable_s2t_host=localhost
+--ttable_t2s_server_port=9494
+--ttable_t2s_host=localhost
+--ttable_server_template=giza_ibm_model1_filtered/genres/$GENRE/align/$DIRECTION/$DIRECTION.mode1.final.gz
+--max_source_phrase=9
+--max_source_elements=5
+--max_terminal_length=5
+--max_nonterminal_span=10
+--hr_max_height=10
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/test/resources/TestConfigFile]: For testing purposes you could lower these parameters.

File: java/ruleXtract/src/test/resources/TestConfigFile [hunk 32]
```diff
@@ -0,0 +1,33 @@
...
+--ttable_t2s_server_port=9494
+--ttable_t2s_host=localhost
+--retrieval_threads=10
+--hfile=hfile
+--test_file=RU/RU.tune.idx
+--rules=G/rules.RU.tune.idx.gz
+--min_source2target_phrase=0.01
+--min_target2source_phrase=1e-10
+--min_source2target_rule=0.01
+--min_target2source_rule=1e-10
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/test/resources/TestConfigFile]: and these ones to 0.01

File: java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/RuleSpec.scala [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+
+package uk.ac.cam.eng.extraction
+
+import org.scalatest._
+import java.io.ByteArrayOutputStream
+import java.io.DataOutputStream
+import java.io.DataInputStream
+import java.io.ByteArrayInputStream
+
+
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/RuleSpec.scala]: Please document. Is this a test? The name of the class could be RuleWhateverTest

File: java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/ExtractSpec.scala [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *******************************************************************************/
+
+package uk.ac.cam.eng.extraction
+
+import org.scalatest._
+
+class ExtractSpec extends FlatSpec with Matchers {
```
Comment (gonzaloiglesiasiglesias) [inline java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/ExtractSpec.scala]: Please document the test. The class name should be more meaningful. RuleExtractTest?

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java (hunk 14 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java]: This a comment attached to existing code. I haven't edited it. I agree that that it would be good to have constants but this is out of scope of the changes.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/phrase/PhraseJob.java (hunk 16 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/features/phrase/PhraseJob.java]: This is a refactor of existing code. Again, the return codes were added by Juan, and out of scope of this change.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/CopyRecordsForTesting.java (hunk 18 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/CopyRecordsForTesting.java]: I'm using the standard eclipse indention style, which encourages narrow lines. I don't really want to mess around with custom eclipse styles, and would encourage other people to use the standard eclipse styles.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java (hunk 19 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java]: I think in this case it's obvious that it's a block size of 64k

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java (hunk 20 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/util/SimpleHFileOutputFormat.java]: I think it's reasonably obvious that it is not meant to be a meaningful value. This is an HBase method call and they have not provided a constant. I don't want to start documenting the HBase api. If you use eclipse then you can get a description of the method arguments by hovering over the method with your cursor.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java (hunk 22 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/filtering/RuleFilter.java]: Again, this is due to the Eclipse style.

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/HFileRuleQuery.java (hunk 25 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/rule/retrieval/HFileRuleQuery.java]: Eclipe...

File: java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala (hunk 30 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/scala/uk/ac/cam/eng/extraction/Extract.scala]: I'm confused by this comment. Which line are you referring to specifically?

File: java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/RuleSpec.scala (hunk 33 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/RuleSpec.scala]: Please see:

http://www.scalatest.org/user_guide/selecting_a_style

File: java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/ExtractSpec.scala (hunk 34 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/test/scala/uk/ac/cam/eng/extraction/ExtractSpec.scala]: Again, please see

http://www.scalatest.org/user_guide/selecting_a_style

File: java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java (hunk 13 above)
Comment (aurelienwaite) [inline java/ruleXtract/src/main/java/uk/ac/cam/eng/extraction/hadoop/extraction/ExtractorJob.java]: No, because One is of type IntWritable and putAlignmentCount takes a second argument of type int, which is a Java primitive.
